app/bundles/CoreBundle/Assets/css/app/less/components/buttons.less

```diff
@@ -61,26 +88,251 @@
     &.btn-danger { color: @brand-danger; }
     &.btn-info { color: @brand-info; }
   }
+
+  &.btn-block {width: 100%;}
+}
+
+
+// Hierarchycal buttons
+// ---------------------------------
+
+/// Primary
+.btn-primary, .btn-primary + .dropdown-toggle {
+  background-color: var(--button-primary);
+  color: var(--text-on-color);
+  i {color: var(--text-on-color);}
+  &:hover {background-color: var(--button-primary-hover);}
+  &:active {background-color: var(--button-primary-active);}
+  &:focus {
+    outline-color: var(--focus);
+    box-shadow: inset 0 0 0 1px var(--focus),inset 0 0 0 2px var(--background,#ffffff);
+  }
+  &:disabled {
+    background-color: var(--button-disabled);
+    color: var(--text-on-color-disabled);
+  }
+
+  /// Primary danger
+  &.btn-danger {
```

Comment:
2 problems here (and in the secondary button styles):

By nesting the danger button like this in Sass, you create some concatenation of classes in the CSS output that are probably not what you intended:  
```
.btn-primary.btn-danger,
.btn-primary + .dropdown-toggle.btn-danger {
```
I think you were expecting the `btn-danger` class to attach to the primary button instead of the toggle?

Other problem, it that you don't need nesting (I think). It's one of those cases where nesting doesn't seem needed for the CSS to function so you should avoid it.

Eg. `.btn-danger, .btn-danger + .dropdown-toggle {`

Reply:
hm, I believe the nesting is needed because the danger modifier for a primary button is different from the danger for a tertiary button, and when using danger you also have to specify primary or tertiary to properly have an effect

we have two possibilities here: `btn btn-tertiary btn-danger` and `btn btn-primary btn-danger` with completely different styling

I'll check the issue regarding the dropdown-toggle

Reply:
Okay, so `btn-danger` is a modifier class and so there's different styles to override depending on the type of button, I get it now. 
That said, there is a code-convention for writing modifiers vs variants: using prefixes. The class could be `.is-danger` instead of `btn-danger`. It's not used a lot yet but existing examples in mautic are: `is-prerelease` and `is-not-allowed`.

Reply:
I'm using as modifier to reduce code
but I could, for example, duplicate the main styles for primary, tertiary and ghost to have each as variants

Since we have danger classes for years now, devs are probably used to work this way in Mautic

Will it be good to change that now?

Reply:
-  if you want to keep it used as a modifier, I would just suggest changing the class name to `.is-danger`
- if you want/need to keep the class name as `btn-danger`, they yeah, you'd need to make duplicate stylings to create variants.
In the end, the choice is up to you but just try to use a code convention to make it easier to iterate and build on later on.

Reply:
The convention would totally make sense for an UI built from scratch

I'm just thinking now about other point: we need to maximize compatibility with the current framework (and with already existing btn-danger buttons that would be affected)

Integrate variants in this PR could cause UI bugs with default btn-danger buttons, but they could be fixed in the other where I replace all btn-default by btn-ghost (then btn-danger could be replaced by btn-ghost-danger, but what if not all of them are btn-ghost? Will we be able to check as a user every replacement to ensure it's working as expected?)

I feel that the development complexity level would increase. How much different is the current Bootstrap solution from a modifier instead of variation? Are we overthinking about this?

Reply:
Overthinking? Maybe a little bit :D  But it's okay to try and think ahead a little bit.
Alright, if this train of thought is blocking you from moving forward, it's perfectly okay to park some of it.
Maybe just do your implementation with nesting and keep the idea of prefixed modifier classes for later cleanup.
Eg. can make sense to dig deeper into this when moving to 6.x, as part of getting rid of current framework stuff.

Reply:
The way I'm working considers that we'll keep using Bootstrap as base (but it needs to be updated, _really_)
it's more feasible to move Mautic forward with Bootstrap, having to handle a removal would be a waste of time at this point

Reply:
@LordRembo this is an open conversation, are you satisfied with this PR in its current state for merging, or is more work needed? Thanks!

Reply:
Only left open so he could see the answer, but he's satisfied